Prevent extra refreshes in pkg states

DemiMarie opened the pull request:
Salt uses a file (called “rtag”) to indicate if the package database
needs to be refreshed.  Remove it if we are going to refresh the cache
with `--clean`.

Fixes QubesOS/qubes-issues#6332

marmarek commented:
Please keep the changes to absolutely minimum - specifically do not remove compat code. And mark qubes-specific changes with "QUBES-DOM0" comment. This greatly simplifies syncing with original upstream version (yumpkg.py)

marmarek commented on _modules/qubes_dom0_update.py:
```diff
@@ -489,9 +489,11 @@ def latest_version(*names, **kwargs):
+        # Remove rtag file to keep multiple refreshes from happening in pkg states
+        salt.utils.pkg.clear_rtag(__opts__)
+        cmd.append('--clean')
+    cmd.extend(['--action=list', '--available'])
```
This makes it behave "funny" with `--clean`. Specifically, the VM's script will call `dnf --available clean all` ([here](https://github.com/QubesOS/qubes-core-agent-linux/blob/master/package-managers/qubes-download-dom0-updates.sh#L87)). Fortunately, that is before `set -e`, so it doesn't abort the whole thing, but it makes output confusing. That said, it seems to be rather a bug in the VM's script.
Same applies to other list options.

DemiMarie replied:
> This doesn't solve the issue. Calling `qubesctl state.sls update.qubes-dom0` still makes those calls:
> 
> * `qubes-dom0-update --quiet --clean --action=list --upgrades`
> * `qubes-dom0-update --quiet --assumeyes --clean --action=clean expire-cache`
> * `qubes-dom0-update --quiet --assumeyes --clean --action=check-update`
> * `qubes-dom0-update --quiet -y --clean --action=upgrade`
> 
> But if I remove `refresh: True` from the `update.qubes-dom0` state, then `--clean` is not used in any of the calls. And the two middle calls don't happen at all. Theoretically dnf should automatically refresh outdated cache, but we hit bugs related to this code in the past (like https://bugzilla.redhat.com/1669247).

Solution: refresh the cache ourselves, but only if we have not done so in this state run.  Force-push upcoming.

marmarek reviewed (changes requested):
This doesn't solve the issue. Calling `qubesctl state.sls update.qubes-dom0` still makes those calls:
- `qubes-dom0-update --quiet --clean --action=list --upgrades`
- `qubes-dom0-update --quiet --assumeyes --clean --action=clean expire-cache`
- `qubes-dom0-update --quiet --assumeyes --clean --action=check-update`
- `qubes-dom0-update --quiet -y --clean --action=upgrade`

But if I remove `refresh: True` from the `update.qubes-dom0` state, then `--clean` is not used in any of the calls. And the two middle calls don't happen at all. Theoretically dnf should automatically refresh outdated cache, but we hit bugs related to this code in the past (like https://bugzilla.redhat.com/1669247).

marmarek commented:
This version still behaves like the previous one: with no `refresh: True`, no call gets `--clean`.
Looking at the salt code, `rtag` is created [only when calling `pkg.installed` or `pkg.latest`](https://github.com/saltstack/salt/blob/master/salt/states/pkg.py#L3476). So, it isn't surprising that using rtag makes no call getting `--clean` when using update-related functions only.

marmarek commented on _modules/qubes_dom0_update.py:
```diff
@@ -3311,3 +3319,109 @@ def list_installed_patches():
     '''
     return _get_patches(installed_only=True)
+
+def mod_init(low):
```
According to this very comment copied from `salt.states.pkg`, it should be named `ex_mod_init`.

DemiMarie replied:
Whoops!  Sorry about that.

marmarek commented on _modules/qubes_dom0_update.py:
```diff
@@ -3311,3 +3319,109 @@ def list_installed_patches():
+    return False
+
+
+def mod_aggregate(low, chunks, running):
```
This, and the `mod_watch` I think are not really needed there. They are done at the `salt.states.pkg` level (this file provides `salt.modules.pkg`).

DemiMarie replied:
Agreed.

marmarek commented on _modules/qubes_dom0_update.py:
```diff
@@ -1032,8 +1032,11 @@ def list_upgrades(refresh=True, **kwargs):
+        # Remove rtag file to keep multiple refreshes from happening in pkg states
+        salt.utils.pkg.clear_rtag(__opts__)
+        cmd.append('--clean')
+    cmd.extend(['--action=list', '--upgrades'])
```
Rebase error? This is now twice.

DemiMarie replied:
Oops!  Will fix.

marmarek replied:
This is still the case.

marmarek commented on update/qubes-dom0.sls:
```diff
@@ -11,5 +11,4 @@
 update:
-  pkg.uptodate:
-    - refresh: True
+  pkg.uptodate: []
```
Sadly, this also isn't enough. It seems [pkg.uptodate state supports only refresh `True/False`](https://github.com/saltstack/salt/blob/master/salt/states/pkg.py#L3263), specifically not `None` which would make salt using rtag to decide. If it's set to False (the default), refreshing metadata is skipped, regardless of the rtag.
Maybe it's time for explicit metadata refresh in a separate state here?

DemiMarie replied:
Darn :disappointed:.  Explicitly refreshing the metadata is probably the way to go here.

DemiMarie committed:
Avoid extra refreshes when updating dom0

Salt doesn’t support this well at all.  As a workaround, first refresh
the cache manually, then tell Salt further refreshes are not needed.

Fixes QubesOS/qubes-issues#6332
